### PR TITLE
SY-4159: Enable Backtick Raw String Literals in Arc

### PR DESCRIPTION
**Pull request by sy-nico**
# Issue Pull Request

## Linear Issue

[SY-4159](https://linear.app/synnax/issue/SY-4159)

## Description

Adds Go-style backtick raw string literals to Arc. A backtick-delimited string takes its body verbatim from source: every byte between the opening and closing backtick (including newlines, tabs, and backslashes) becomes part of the runtime string with no escape interpretation.

```go
x := `hello`
y := `a
    b`
z := `say "hi"`
```

<h3>Greptile Summary</h3>

This PR adds Go-style backtick raw string literals to Arc. Unlike Go's raw strings (which cannot contain backticks at all), Arc's variant supports a single `\`` escape to embed a literal backtick while keeping all other bytes verbatim — including real newlines, tabs, and backslashes.

- **Lexer/parser**: A new `STR_LITERAL_RAW` token is defined in the ANTLR grammar and wired into the `literal` parser rule, with the generated Go parser and interpreter files regenerated accordingly.
- **Runtime support**: `ParseRawString` and `unescapeRawBackticks` are added to `arc/go/literal`; the compiler, type-inferrer, and formatter all route the new token through dedicated handlers.
- **LSP / editor integration**: `appendTokenPerLine` is introduced in `semantic.go` to split any multi-line token into per-line LSP entries (fixing token coloring for multi-line raw strings in Monaco); a new `stringRaw` semantic token type and corresponding TextMate grammar rule provide distinct syntax highlighting in both the Go LSP server and the TypeScript console.

<h3>Confidence Score: 5/5</h3>

Safe to merge — all changed paths are additive and well-tested with no regressions on existing string handling.

The change is purely additive: a new token kind, a new parse function, and a new LSP handler. Existing STR_LITERAL paths are untouched. The appendTokenPerLine refactor is behaviorally equivalent for single-line tokens and correctly handles multi-line splits. Test coverage spans the lexer, parser, type inferrer, compiler, formatter, and LSP layers. The two findings are minor style/robustness observations in test and utility code.

No files require special attention, though semantic_test.go contains hardcoded numeric iota values that should track semantic.go manually.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| arc/go/parser/ArcLexer.g4 | Adds STR_LITERAL_RAW token rule; grammar correctly handles \` as the only escape inside backtick delimiters. |
| arc/go/parser/ArcParser.g4 | Adds STR_LITERAL_RAW as an alternative in the literal rule; change is minimal and correct. |
| arc/go/literal/literal.go | Adds ParseRawString and unescapeRawBackticks; validation logic is correct but unescapeRawBackticks allocates unnecessarily for the escape-free common case. |
| arc/go/lsp/semantic.go | Refactors per-token append into appendTokenPerLine that correctly splits multi-line raw strings; also benefits pre-existing multi-line comments. Logic is sound. |
| arc/go/lsp/semantic_test.go | New test file with good coverage of single/multi-line token splitting; token-type constants are hardcoded numeric indices that will silently drift if the iota ordering changes. |
| arc/go/compiler/expression/literal.go | Routes STR_LITERAL_RAW tokens through ParseRawString via the generalized compileStringLiteral signature; clean and correct. |
| arc/go/analyzer/types/infer_expression.go | Replaces fragile text[0]-based heuristic with proper AST node presence checks for both string literal kinds. |
| console/src/arc/lsp/lsp.ts | Registers the stringRaw semantic token type with theme colors; colors are intentionally distinct from regular strings. |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["Arc source: \`...\`"] --> B[ANTLR STR_LITERAL_RAW token]
    B --> C{consumer}
    C -->|compiler| D["compileLiteral → ParseRawString"]
    C -->|type inferrer| E["STR_LITERAL_RAW() != nil → String type"]
    C -->|formatter| F["isLiteral() → preserve verbatim"]
    C -->|LSP server| G["mapLexerTokenType → SemanticTokenTypeStringRaw"]
    G --> H["appendTokenPerLine — split on newlines"]
    H --> I["Per-line LSP Token entries"]
    D --> J["unescapeRawBackticks — only \\` → backtick"]
    J --> K["ParsedValue{Value: body}"]
```

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `arc/go/lsp/semantic.go`, line 292-309 ([link](https://github.com/synnaxlabs/synnax/blob/e587b3ee6da5780c8bcc5e9264ee6cef1fd0c0c0/arc/go/lsp/semantic.go#L292-L309)) 

   <a href="#">P2</a> **Zero-length semantic tokens on consecutive newlines**

   When a raw string literal contains two adjacent newlines (e.g., `` `a\n\nb` ``), the loop emits a token with `Length: 0` for the empty middle line (`i - lineStart == 0`). The LSP specification does not prohibit zero-length tokens, but Monaco may behave unexpectedly — in practice it is safer to skip emitting a token for an empty segment. No existing test covers consecutive-newline input, so this edge case is currently unverified.
</details>

<sub>Reviews (3): Last reviewed commit: ["SY-4159: Handle backticks with \\\` in a \`..."](https://github.com/synnaxlabs/synnax/commit/ad057b8167ce1ec28cb67d38062d18d3b43f3e32) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=31255548)</sub>

### CODE REVIEW

File: arc/go/parser/ArcLexer.g4 [hunk 1]
```diff
@@ -116,6 +116,11 @@ STR_LITERAL
     : '"' (~["\\\r\n] | ESCAPE_SEQUENCE)* '"'
     ;
 
+// Raw multi-line string literal.
+STR_LITERAL_RAW
+    : '`' ~[`]* '`'
+    ;
```
Comment (pjdotson) [inline arc/go/parser/ArcLexer.g4:123]: I just want to make sure you are aware, this grammar means that a raw string literal
CANNOT have a backtick in it (this is the behavior in Go). In JavaScript, you can
include an escaped backtick with `\``. Is this the behavior you want?

**Review by pjdotson — CHANGES_REQUESTED** [review]

File: arc/go/parser/ArcLexer.g4 (hunk 1 above)
Comment (sy-nico) [reply arc/go/parser/ArcLexer.g4:123]: I do want to be able to have backticks. JS uses \` (which maybe you typed backtick, slash, backtick, backtick -> `\``
 I will correct.
image

Comment (sy-nico) [reply arc/go/parser/ArcLexer.g4:123]: Need to enable `\` with `\\`
